client/post-editor/editor-publish-button/index.jsx

```diff
@@ -0,0 +1,127 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { recordEvent } from 'lib/posts/stats';
+import postUtils from 'lib/posts/utils';
+import siteUtils from 'lib/site/utils';
+import Button from 'components/button';
+
+export const getPublishButtonStatus = ( site, post, savedPost ) => {
+	if (
+		postUtils.isPublished( savedPost ) &&
+		! postUtils.isBackDatedPublished( savedPost ) &&
+		! postUtils.isFutureDated( post ) ||
+		(
+			savedPost &&
+			savedPost.status === 'future' &&
+			postUtils.isFutureDated( post )
+		)
+	) {
+		return 'update';
+	}
+
+	if ( postUtils.isFutureDated( post ) ) {
+		return 'schedule';
+	}
+
+	if ( siteUtils.userCan( 'publish_posts', site ) ) {
+		return 'publish';
+	}
+
+	if ( savedPost && savedPost.status === 'pending' ) {
+		return 'update';
+	}
+
+	return 'requestReview';
+};
+
+export default React.createClass( {
```

Comment:
A few recommendations on refactoring, perhaps best for a later PR:

- React core meeting notes highlight that `React.createClass` is on a path toward deprecation, so we should use either [stateless function components](https://facebook.github.io/react/docs/reusable-components.html#stateless-functions) or the `Component` class
- In the path toward Redux, we should seek to have components derive as much from global state as possible using `react-redux` `connect` with simplest props possible. This is in contrast with previous convention to pass very large objects through large hierarchy of components. See also: ["Data Components"](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#data-components)
- We're moving toward recommending all new components be used with the [`i18n-calypso` `localize` higher order component](https://github.com/Automattic/i18n-calypso#localize) in favor of the `this.translate` mixin ([soon to be a warning](https://github.com/Automattic/eslint-plugin-wpcalypso/pull/28))